v2/api_error_test.go

```diff
@@ -0,0 +1,195 @@
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//     * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+package gax
+
+import (
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"google.golang.org/genproto/googleapis/rpc/errdetails"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/durationpb"
+)
+
+func TestDetails(t *testing.T) {
+	qf := &errdetails.QuotaFailure{
+		Violations: []*errdetails.QuotaFailure_Violation{{Subject: "Foo", Description: "Bar"}},
+	}
+	st, _ := status.New(codes.ResourceExhausted, "qf").WithDetails(qf)
+	apierr := &APIError{
+		err:     st.Err(),
+		status:  st,
+		details: ErrDetails{QuotaFailure: qf},
+	}
+	if diff := cmp.Diff(ErrDetails{QuotaFailure: qf}, apierr.Details(), cmp.Comparer(proto.Equal)); diff != "" {
+		t.Errorf("Expected(+), Actual(-):\n%s", diff)
+	}
+}
+
+func TestError(t *testing.T) {
+	pf := &errdetails.PreconditionFailure{
+		Violations: []*errdetails.PreconditionFailure_Violation{{Type: "Foo", Subject: "Bar", Description: "test"}},
+	}
+	st, _ := status.New(codes.FailedPrecondition, "pf").WithDetails(pf)
+	apierr := &APIError{
+		err:     st.Err(),
+		status:  st,
+		details: ErrDetails{PreconditionFailure: pf},
+	}
+	if !strings.Contains(apierr.Error(), "Foo") {
+		t.Errorf("Status message not included!")
+	}
+}
+
+func TestGRPCStatus(t *testing.T) {
+	qf := &errdetails.QuotaFailure{
+		Violations: []*errdetails.QuotaFailure_Violation{{Subject: "Foo", Description: "Bar"}},
+	}
+	st, _ := status.New(codes.ResourceExhausted, "Per user quota has been exhausted").WithDetails(qf)
+	apierr := &APIError{
+		err:     st.Err(),
+		status:  st,
+		details: ErrDetails{QuotaFailure: qf},
+	}
+	if st != apierr.GRPCStatus() {
+		t.Errorf("Expected: %v but got: %v", st, apierr.GRPCStatus())
```

Comment:
Unlike some other languages like Java we generally put the got first then the want.